leapsecs_read.c

```diff
@@ -0,0 +1,50 @@
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <errno.h>
+extern int errno;
+#include "tai.h"
+#include "leapsecs.h"
+
+struct tai *leapsecs = 0;
+int leapsecs_num = 0;
+
+int leapsecs_read()
+{
+  int fd;
+  struct stat st;
+  struct tai *t;
+  int n;
+  int i;
+  struct tai u;
+
+  fd = open("/etc/leapsecs.dat",O_RDONLY | O_NDELAY);
+  if (fd == -1) {
+    if (errno != ENOENT) return -1;
+    if (leapsecs) free(leapsecs);
+    leapsecs = 0;
+    leapsecs_num = 0;
+    return 0;
+  }
+
+  if (fstat(fd,&st) == -1) { close(fd); return -1; }
+
+  t = (struct tai *) malloc(st.st_size);
```

Comment:
No, not in any way. If I do just ```echo 0 > /etc/leapsecs.dat" than this will just explode. Better would be to first check that the file size is sensible, and more generally do:

```
t = malloc(sizeof(*t));
```

But I see t neither freed nor returned, so this is probably leaking? And if it is just a local variable, why not just declare it on the stack and scrap the whole heap allocation? This would still need the sanity check on the file size, but removes at least one different error case.

Reply:
This does not explode. In circumstances where /etc/leapsecs.dat is less than the size of a struct t, n is rounded down to 0 and the *leapsecs pointer is assigned to an empty t the size of st.st_size.

Reply:
Yes, you are right, but I still find it dangerous, and also unnecessary.

 * if it's smaller than struct tai one can immediately leave the function and not do any allocation at all as it would be ignored anyway
 * if it's bigger than struct tai it is completely read into memory, even if it is _big_, so it could DoS you if you drop a wrong file there by accident. I would mmap() it if one really needs the whole content.

Reply:
This also does no need a cast if the malloc() would be kept, as void* can be assigned to any other pointer type in C without cast. May come from a time where it was alloc(), which returned char*.